src/USER-HSMA/hsma2d.cpp

```diff
@@ -0,0 +1,2309 @@
+					IntegralDown[m * Nw + i][n * Nw + j][0] = ar[m][i] * Lx / 2.0;
+					IntegralDown[m * Nw + i][n * Nw + j][1] = ar[n][j] * Ly / 2.0;
+					IntegralDown[m * Nw + i][n * Nw + j][2] = -Lz / 2;
+					IntegralDown[m * Nw + i][n * Nw + j][3] = Weight[i] * Weight[j] * Lx * Ly / (4.0 * S * S);
+				}
+	
+	Rs = Lambda * sqrt((Lx / 2) * (Lx / 2) + (Ly / 2) * (Ly / 2) + (Lz / 2) * (Lz / 2));
+	PI = 3.141592653589793;
+	mu = 2.0 * PI / Lx;
+	EU = (1 - Gamma) / (1 + Gamma);
+
+	IndexReal = floor((p * p - 1) / 2.0);
+	IndexImag = ceil((p * p - 1) / 2.0);
+	NJKBound = floor(2.0 * p / 3.0);
+	NJK = (2 * NJKBound + 1) * (2 * NJKBound + 1);
+	LeftTermReal = new double* [2 * NJK];
+	LeftTermImag = new double* [2 * NJK];
+	for (int i = 0; i < 2 * NJK; i++)
+	{
+		LeftTermReal[i] = new double[IndexReal];
+		LeftTermImag[i] = new double[IndexImag];
+	}
+	for (int i = 0; i < 2 * NJK; i++)
+	{
+		for (int j = 0; j < IndexReal; j++)
+		{
+			LeftTermReal[i][j] = 0.00;
+		}
+		for (int j = 0; j < IndexImag; j++)
+		{
+			LeftTermImag[i][j] = 0.00;
+		}
+	}
+	ConstructLeftTerm((double**)LeftTermReal, (double**)LeftTermImag, (double***)IntegralTop, (double***)IntegralDown, Nw, NJKBound, p, EU, mu, Lx, Ly, Lz, S);//Construct the left term.
+
+	 //Construct Fibonacci Points And There Local/Multipole Expansion.  Fibonacci[Np][4] storage the positions and the weights.
+	Np = int(2 * F + 2.0);
+	Fibonacci = new double* [Np];
+	for (int i = 0; i < Np; i++)
+	{
+		Fibonacci[i] = new double[4];
+	}
+	QRD = new double* [Np]; QLocalRD = new double* [Np];
+	for (int i = 0; i < Np; i++)
+	{
+		QRD[i] = new double[p * p];
+		QLocalRD[i]= new double[p * p];
+	}
+	SetFibonacci(Fibonacci, F, Fp, Np, Rs, PI);
+    #pragma omp parallel
+	{
+		double MulQ[p * p], MulLocalQ[p * p];
+        #pragma omp for schedule(static) private(MulQ,MulLocalQ)
+		for (int i = 0; i < Np; i++)
+		{
+			CalculateRDMultipoleExpansion(MulQ, p, Fibonacci[i][0], Fibonacci[i][1], Fibonacci[i][2]);
+			CalculateLocalRDMultipoleExpansion(MulLocalQ, p, Fibonacci[i][0], Fibonacci[i][1], Fibonacci[i][2], Rs);
+			for (int j = 0; j < p * p; j++)
+			{
+				QRD[i][j] = MulQ[j];
+				QLocalRD[i][j] = MulLocalQ[j];
+			}
+		}
+	}
+
+	TopNear = new double* [S * Nw];
+	TopZDNear = new double* [S * Nw];
+	DownNear = new double* [S * Nw];
+	DownZDNear = new double* [S * Nw];
+	for (int i = 0; i < S * Nw; i++)
+	{
+		TopNear[i] = new double[S * Nw];
+		TopZDNear[i] = new double[S * Nw];
+		DownNear[i] = new double[S * Nw];
+		DownZDNear[i] = new double[S * Nw];
+	}
+
+	AR = new double[S * Nw];
+	for (int i = 0; i < S; i++)
+	{
+		for (int j = 0; j < Nw; j++)
+		{
+			AR[i * Nw + j] = ar[i][j];
+		}
+	}
+
+	cout << Lx << "=Lx   " << Ly << "=Ly    " << Lz << "=Lz   " << Lambda << "=Lambda    " <<Gamma<<"=Gamma    "<<w<<"=w    "<< p << "=p   " << Nw << "=Nw    " << Fp << "=Fp   " << F << "=F   " << IF_FMM_RightTerm << "=IF_FMM_RightTerm   " << IF_FMM_FinalPotential << "=IF_FMM_FinalPotential   " << endl;
```

Comment:
Please do not mix stdio and C++ iostreams. LAMMPS has settled for using C style stdio throughout. And see my comment from above about using utils::logmesg() for informational messages. However, this looks more like a leftover debug message and should probably removed alltogether.

Reply:
Revised